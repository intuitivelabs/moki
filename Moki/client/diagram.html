<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <link rel="shortcut icon" id="favicon" href="/favicon.ico">

  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="theme-color" content="#000000">

  <title>Monitor</title>
</head>

<body>
  <noscript>You need to enable JavaScript to run this app.</noscript>
  <sequence-diagram />

  <script type=module>
    import "bootstrap/dist/css/bootstrap.min.css";
    import "bootstrap/dist/js/bootstrap.min.js";
    import "./src/styles/style.scss";
    import "./src/gui/src/style/style.css";

    document.body.style.setProperty('--main', "#58a959");
    document.body.style.setProperty('--second', "#58a959");
  </script>
  <script type="module" src="/src/js/diagram/DiagramWebComponent.tsx"></script>
  <script>
    // D-ENTRYPOINT
    const data = ``;
    window.addEventListener("load", (event) => {
      const diagram = document.querySelector('sequence-diagram');
      diagram.content = data;
    });
  </script>
</body>

</html>