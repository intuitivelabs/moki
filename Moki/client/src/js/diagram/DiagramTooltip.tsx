import { CSSProperties, MutableRefObject, SetStateAction, useEffect, useRef } from "react";
import { IconX } from "@tabler/icons-react";

function useDraggable(
  windowRef: React.RefObject<HTMLElement>,
  dragRef: React.RefObject<HTMLElement>,
  { offsetX = 0, offsetY = 0, deps }: {
    offsetX?: number;
    offsetY?: number;
    deps: React.DependencyList;
  },
) {
  const offset = useRef({ x: 0, y: 0 });

  const handleDown = (e: MouseEvent) => {
    const element = windowRef.current;
    if (!element) return;
    e = e || window.event;
    e.preventDefault();

    const rect = windowRef.current.getBoundingClientRect();
    offset.current.x = e.clientX - rect.x + offsetX;
    offset.current.y = e.clientY - rect.y + offsetY;
    document.onmouseup = handleUp;
    document.onmousemove = handleMove;
  };

  const handleMove = (e: MouseEvent) => {
    const element = windowRef.current;
    if (!element) return;
    e = e || window.event;
    e.preventDefault();

    const x = e.pageX - offset.current.x;
    const y = e.pageY - offset.current.y;
    element.style.transform = `translate(${x}px, ${y}px)`;
  };

  const handleUp = () => {
    document.onmouseup = null;
    document.onmousemove = null;
  };

  useEffect(() => {
    const element = dragRef.current;
    if (!element) return;
    element.onmousedown = handleDown;
  }, [dragRef, offsetX, offsetY, ...deps]);
}

interface Props {
  isOpen: boolean;
  setIsOpen: (value: SetStateAction<boolean>) => void;
  floatingStyles: React.CSSProperties;
  pushFront: () => void;
  headerStyle: CSSProperties;
  headerContent: JSX.Element;
  bodyContent: JSX.Element;
  refs: {
    floating: MutableRefObject<HTMLElement | null>;
    setFloating: (node: HTMLElement | null) => void;
  };
}

function DiagramTooltip(
  {
    isOpen,
    setIsOpen,
    floatingStyles,
    refs,
    pushFront,
    headerStyle,
    headerContent,
    bodyContent,
  }: Props,
) {
  const dragRef = useRef<HTMLDivElement>(null);
  useDraggable(refs.floating, dragRef, {
    deps: [isOpen],
  });

  return (
    <>
      {isOpen && (
        <div
          className="tooltipDiagram"
          ref={refs.setFloating}
          style={floatingStyles}
          onMouseDown={pushFront}
        >
          <div
            ref={dragRef}
            style={{
              cursor: "move",
              display: "flex",
              alignItems: "center",
              ...headerStyle,
            }}
          >
            <button
              className="border-0 px-2"
              style={{ background: "none", paddingBottom: "0.3rem" }}
              onClick={() => {
                setIsOpen(false);
              }}
            >
              <IconX size={19} />
            </button>
            {headerContent}
          </div>
          {bodyContent}
        </div>
      )}
    </>
  );
}

export { DiagramTooltip };
