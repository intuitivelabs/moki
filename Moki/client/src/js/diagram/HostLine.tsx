import { Line } from "@visx/shape";
import { Text } from "@visx/text";
import { Group } from "@visx/group";
import { Fragment } from "react";

interface Props {
  activeHosts: string[];
  classWidth: number;
  vertSpace: number;
  xPad: number;
  height: number;
}

const VERT_PAD = 20;
const YPAD = 30;

function HostLines(
  { activeHosts, classWidth, height, vertSpace, xPad }: Props,
) {
  return (
    <Group>
      {activeHosts.map((name, i) => {
        const x = xPad + i * vertSpace;
        return (
          <Fragment key={i}>
            <Line
              style={{ stroke: "#888", strokeDasharray: "3, 3" }}
              from={{ x, y: YPAD + 20 }}
              to={{ x, y: YPAD + VERT_PAD + height }}
            />
            <Group
              transform={`translate(${x}, ${YPAD})`}
              className="class-rect"
            >
              <rect
                x={-classWidth / 2}
                width={classWidth}
                height={24}
                opacity={0.1}
                rx={5}
              />
            </Group>
            <Group transform={`translate(${x}, ${YPAD})`}>
              <Text className="class-label" textAnchor="middle" dy={16}>
                {name}
              </Text>
            </Group>
          </Fragment>
        );
      })}
    </Group>
  );
}

export { HostLines };
