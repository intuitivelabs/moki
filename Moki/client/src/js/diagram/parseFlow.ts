import { formatDate } from "../helpers/formatTime";

interface Flow {
  sipFlow: Pick<FlowSIP, "sips" | "hosts" | "branches">;
  syslogs: PacketSyslog[];
}

interface FlowSIP {
  sips: PacketSIP[];
  hosts: Set<string>;
  branches: Set<string>;
  idCounter: number;
  lastTimestamp: Date | undefined;
}

// timestep in milliseconds
interface PacketSIP {
  id: number;
  method: string;
  timestep: number;
  branch: string;
  src: string;
  dst: string;
  timestamp: string;
  payload: string;
}

interface PacketSyslog {
  facility: string;
  severity: string;
  timestamp: string;
  payload: string;
}

// utilities

function getTimestamp(node: Element): Date | null {
  const timestamp = node.getAttribute("timestamp");
  if (timestamp == null) return null;
  const parsedTimestamp = new Date(timestamp);
  if (parsedTimestamp == null) return null;
  return parsedTimestamp;
}

// Parse XML flow following decap format
// Can throw an error
function parseFlow(flowDataXML: string): Flow {
  const parser = new DOMParser();
  const doc = parser.parseFromString(flowDataXML, "text/xml");

  if (doc.documentElement.nodeName == "parsererror") {
    throw new Error(doc.documentElement.textContent ?? "parse error");
  }

  const sipFlow: FlowSIP = {
    sips: [],
    hosts: new Set(),
    branches: new Set(),
    lastTimestamp: undefined,
    idCounter: 0,
  };
  const syslogs = [];

  const flow = doc.documentElement;
  for (const node of flow.children) {
    switch (node.nodeName) {
      case "sip": {
        const packet = parseSIP(node, sipFlow);
        if (packet != null) sipFlow.sips.push(packet);
        break;
      }
      case "syslog": {
        const packet = parseSyslog(node);
        if (packet != null) syslogs.push(packet);
        break;
      }
    }
  }

  return { sipFlow, syslogs };
}

function parseSIP(node: Element, flow: FlowSIP): PacketSIP | null {
  const src = node.getAttribute("src");
  const dst = node.getAttribute("dst");
  if (!src || !dst) return null;

  // keep tracks of src and dst value
  flow.hosts.add(src);
  flow.hosts.add(dst);

  // keep track of timestamp for time difference between packets
  const parsedTimestamp = getTimestamp(node);
  if (parsedTimestamp == null) return null;

  // first sip packet
  let timestep = 0;
  if (flow.lastTimestamp == undefined) {
    flow.lastTimestamp = parsedTimestamp;
  } else {
    timestep = parsedTimestamp.getTime() -
      flow.lastTimestamp.getTime();
    flow.lastTimestamp = parsedTimestamp;
  }

  const branch = node.getAttribute("branch") ?? "unknown";
  flow.branches.add(branch);
  const id = flow.idCounter;
  flow.idCounter += 1;

  return {
    id,
    timestamp: formatDate(parsedTimestamp.getTime()),
    timestep,
    branch,
    src,
    dst,
    method: node.getAttribute("method") ?? "unknown",
    payload: node.textContent ?? "",
  };
}

function parseSyslog(node: Element): PacketSyslog | null {
  const facility = node.getAttribute("facility") ?? "unknown";
  const severity = node.getAttribute("severity") ?? "unknown";

  const parsedTimestamp = getTimestamp(node);
  if (parsedTimestamp == null) return null;

  return {
    facility,
    severity,
    timestamp: formatDate(parsedTimestamp.getTime()),
    payload: node.textContent ?? "",
  }
}

export type { Flow, PacketSIP };
export { parseFlow };
