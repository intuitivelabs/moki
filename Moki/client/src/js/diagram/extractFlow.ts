import { formatDate, formatSmallDuration } from "../helpers/formatTime";
import { getSeverityLevel } from "./Syslog";
import { Flow, PacketSIP, PacketSyslog } from "./parseFlow";

// SIP affiliated syslog
interface SipSyslog {
  packet: PacketSyslog;
  timestamp: string;
}

interface ExtractedPacket {
  unixTimestamp: number;
  sip: {
    packet: PacketSIP;
    timestamp: string;
  };
  syslog: {
    packets: SipSyslog[];
    severity: string;
  };
}

/**
 * Flow iterator
 * Give back sip packet one by one corresponding syslogs during this timeframe
 */
function* extractFlow(
  flow: Flow,
  sipPackets: PacketSIP[],
  activeHostsSet: Set<string>,
): Generator<ExtractedPacket> {
  const syslogs = flow.syslogs;

  let syslogIndex = 0;
  let lastTimestamp = 0;
  let nextTimestamp: number | undefined = 0;

  for (let i = 0; i < sipPackets.length; i++) {
    const sip = sipPackets[i];

    const extracted: ExtractedPacket = {
      unixTimestamp: sip.unixTimestamp,
      sip: {
        packet: sip,
        timestamp: "unkown",
      },
      syslog: {
        packets: [],
        severity: "",
      },
    };

    nextTimestamp = sipPackets.at(i + 1)?.unixTimestamp;

    // sip timestamp / time diff
    if (lastTimestamp === 0) {
      // first packet
      lastTimestamp = sip.unixTimestamp;
      extracted.sip.timestamp = formatDate(sip.unixTimestamp);
    } else {
      // subsequent packet
      const timeDiff = Math.max(sip.unixTimestamp - lastTimestamp, 0);
      lastTimestamp = sip.unixTimestamp;
      extracted.sip.timestamp = formatSmallDuration(timeDiff);
    }

    // get affiliated syslogs
    const sipSyslogs: SipSyslog[] = [];
    let maxLevel = 0;
    let severity = "unknown";

    while (syslogIndex < syslogs.length) {
      const syslog = syslogs[syslogIndex];
      if (nextTimestamp && syslog.unixTimestamp >= nextTimestamp) {
        break;
      }

      const level = getSeverityLevel(syslog.severity);
      if (level > maxLevel) {
        maxLevel = level;
        severity = syslog.severity;
      }

      const timeDiff = Math.max(syslog.unixTimestamp - lastTimestamp, 0);
      sipSyslogs.push({
        packet: syslog,
        timestamp: formatSmallDuration(timeDiff),
      });

      syslogIndex += 1;
    }

    // populate syslogs
    extracted.syslog.packets = sipSyslogs;
    extracted.syslog.severity = severity;

    yield extracted;
  }
}

export type { ExtractedPacket, SipSyslog };
export { extractFlow };
