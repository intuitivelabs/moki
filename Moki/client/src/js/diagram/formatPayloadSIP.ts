import { PacketSIP } from "./parseFlow";

function escapeHTML(text: string) {
  const element = document.createElement("div");
  const textNode = document.createTextNode(text);
  element.appendChild(textNode);
  return element.innerHTML;
}

function formatSIPPayload(packet: PacketSIP): string {
  // remove first line
  const lines = packet.payload.split("\n");
  lines.shift();
  let formatted = `<div><b>${packet.method}</b></div>`;

  // first part: headers
  for (let i = 0; i < lines.length; i++) {
    const line = lines.shift();
    if (!line || line.length === 0 || line === "\r") {
      lines.shift();
      break;
    }
    // split header name and value
    const index = line.indexOf(":");
    if (index === -1) {
      formatted += `<div><span className="value">${line}</span></div>`;
    } else {
      const [name, value] = [line.slice(0, index), line.slice(index + 1)];
      formatted += `<div>
        <span className="key"><b>${name}: </b></span>
        <span className="value">${escapeHTML(value)}</span>
      </div>`;
    }
  }

  formatted += "<br />";
  // second part: payload
  for (const line of lines) {
    // syntax highlight
    const highlight = /^c=|^m=/.test(line);
    formatted += `<div>
      <span className="value" style=${highlight ? "color:green" : ""}>
        ${line}
      </span></div>`;
  }

  return formatted;
}

export { formatSIPPayload };
