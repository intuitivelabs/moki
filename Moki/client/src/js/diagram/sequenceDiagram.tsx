import React, { useEffect, useRef, useState } from "react";
import {
  curveBasisClosed,
  line,
  ScaleOrdinal,
  scaleOrdinal,
  select,
  selectAll,
} from "d3";

import downloadIcon from "/icons/download.png";
import downloadPcapIcon from "/icons/downloadPcap.png";
import { Flow, parseFlow } from "@/js/diagram/parseFlow";
import { formatSIPPayload } from "@/js/diagram/formatPayloadSIP";
import {
  downloadPcapResource,
  downloadPcaps,
} from "@/js/helpers/download/downloadPcapResource";
import { saveFile } from "@/js/helpers/download/saveFile";

const Tableau10 = [
  "#4e79a7",
  "#f28e2c",
  "#e15759",
  "#76b7b2",
  "#59a14f",
  "#edc949",
  "#af7aa1",
  "#ff9da7",
  "#9c755f",
  "#bab0ab",
];

export default function SequenceDiagram() {
  let filenames: string[] | undefined = undefined;

  // pcap files
  const params = new URLSearchParams(window.location.search);
  const id = params.get("id");
  if (id) {
    filenames = id.split(",");
  }

  /* Load data */
  async function load() {
    if (!filenames || filenames.length === 0) {
      throw new Error("No PCAP filenames specified in id search parameter");
    }

    try {
      const res = await downloadPcapResource(filenames, "api/diagram");
      return await res.text();
    } catch (error) {
      if (error instanceof Error) {
        throw new Error("Problem while receiving data: " + error.message);
      }
    }

    throw new Error("No data");
  }

  return <SequenceDiagramRender {...{ load, filenames }} />;
}

function formatSmallDuration(duration: number) {
  if (duration < 1000) return `+ ${Math.round(duration)}ms`;
  else return `+ ${Math.round(duration / 1000)}s`;
}

interface RenderProps {
  filenames: string[] | undefined;
  load: () => Promise<string>;
  download?: boolean;
}

export function SequenceDiagramRender(
  { load, filenames, download = true }: RenderProps,
) {
  const [data, setData] = useState<Flow | undefined>();
  const [selectedHosts, setSelectedHosts] = useState<Set<string>>();
  const colorScaleSIP = useRef<ScaleOrdinal<string, string>>(scaleOrdinal());
  const [error, setError] = useState<string | undefined>(undefined);

  const noData = !data;

  useEffect(() => {
    const loadData = async () => {
      try {
        const xmlData = await load();
        if (xmlData === "") {
          setError("Error: Empty diagram");
          return;
        }
        const data = parseFlow(xmlData);
        colorScaleSIP.current = scaleOrdinal<string, string>(
          data.sipFlow.branches,
          Tableau10,
        );
        setData(data);
        setSelectedHosts(data.sipFlow.hosts);
      } catch (err) {
        if (err instanceof Error) {
          setError(err.message);
        }
      }
    };
    loadData();
  }, [load]);

  useEffect(() => {
    draw();
  }, [selectedHosts]);

  /**
   * update current selection of node names
   */
  const updateSelection = (
    e: React.MouseEvent<HTMLInputElement, MouseEvent>,
  ) => {
    if (!data) return;

    const id = e.currentTarget.id;
    const checked = e.currentTarget.checked;
    const newSelection = new Set(selectedHosts);

    if (checked) newSelection.add(id);
    else newSelection.delete(id);

    setSelectedHosts(newSelection);
  };

  const getPcap = async () => {
    if (!filenames) return;
    downloadPcaps(filenames, filenames[0]);
  };

  const getDiagram = async () => {
    if (!filenames) return;
    try {
      const res = await downloadPcapResource(
        filenames,
        "api/diagram/bundle",
      );
      const blob = new Blob([await res.blob()], { type: "html" });
      saveFile(blob, filenames[0].slice(0, -5) + ".html");
    } catch (error) {
      if (error instanceof Error) {
        alert("Could not download bundled HTML: " + error.message);
      }
    }
  };

  const draw = () => {
    if (noData || !selectedHosts) return;

    const margin = { top: 0, right: 20, bottom: 20, left: 0 };
    const width = 1200 - (margin.left + margin.right);

    // transform data
    const activeHostsSet = new Set<string>();
    const sipPackets = data.sipFlow.sips.filter((packet) => {
      if (selectedHosts.has(packet.src)) {
        activeHostsSet.add(packet.src);
        return true;
      }
      return false;
    });
    const activeHosts = [...activeHostsSet];

    const chart = document.getElementById("diagramSVG");
    if (chart) {
      chart.remove();
    }

    const svg = select("#diagram")
      .append("svg")
      .attr("width", width + margin.left + margin.right)
      .attr("id", "diagramSVG")
      .attr("transform", `translate(${margin.left}, ${margin.right})`);

    const VERT_SPACE = width / activeHosts.length;
    const VERT_PAD = 20;

    const MESSAGE_SPACE = 30;
    svg.attr("height", (sipPackets.length + 2) * MESSAGE_SPACE + 20);

    const MESSAGE_LABEL_X_OFFSET = -40;
    const MESSAGE_LABEL_Y_OFFSET = 75;
    const MESSAGE_ARROW_Y_OFFSET = 80;

    const CLASS_WIDTH = VERT_SPACE - 10;

    const XPAD = CLASS_WIDTH / 2;
    const YPAD = 30;

    for (const [i, name] of activeHosts.entries()) {
      // Draw vertical lines
      svg.append("line")
        .style("stroke", "#888")
        .style("stroke-dasharray", "3, 3")
        .attr("x1", XPAD + i * VERT_SPACE)
        .attr("y1", YPAD + 20)
        .attr("x2", XPAD + i * VERT_SPACE)
        .attr(
          "y2",
          YPAD + VERT_PAD + sipPackets.length * (MESSAGE_SPACE + 5),
        );

      // Draw classes
      const x = XPAD + i * VERT_SPACE;
      svg.append("g")
        .attr("transform", `translate(${x}, ${YPAD})`)
        .attr("class", "class-rect")
        .append("rect")
        .attr("x", -CLASS_WIDTH / 2)
        .attr("width", CLASS_WIDTH)
        .attr("height", "24px")
        .attr("opacity", 0.1)
        .attr("rx", 5);

      // Draw class labels
      svg.append("g")
        .attr("transform", `translate(${x}, ${YPAD})`)
        .append("text")
        .attr("class", "class-label")
        .attr("text-anchor", "middle")
        .text(name)
        .attr("dy", "16px");
    }

    for (const [i, packet] of sipPackets.entries()) {
      if (!activeHostsSet.has(packet.dst)) continue;

      // draw timestamp labels
      let timestamp = packet.timestamp;
      if (i !== 0) {
        timestamp = formatSmallDuration(packet.timestep);
      }

      svg.append("g")
        .attr(
          "transform",
          `translate(${XPAD + MESSAGE_LABEL_X_OFFSET}, ${
            MESSAGE_LABEL_Y_OFFSET + i * MESSAGE_SPACE
          })`,
        )
        .attr("class", "first")
        .attr("text-anchor", "middle")
        .append("text")
        .style("font-size", "10px")
        .text(timestamp);

      // Draw message arrows
      const arrowY = MESSAGE_ARROW_Y_OFFSET + i * MESSAGE_SPACE;
      const srcIndex = activeHosts.indexOf(packet.src);
      const dstIndex = activeHosts.indexOf(packet.dst);
      const srcArrowX = XPAD + srcIndex * VERT_SPACE;
      const dstArrowX = XPAD + dstIndex * VERT_SPACE;

      if (packet.src === packet.dst) {
        // curved arrow: src -> src
        const pathData = line().curve(curveBasisClosed)([
          [srcArrowX - 10, arrowY - 20],
          [srcArrowX + 10, arrowY + 45],
          [srcArrowX + 45, arrowY + 10],
        ]);
        svg.append("path")
          .attr("d", pathData)
          .attr("stroke", colorScaleSIP.current(packet.branch))
          .attr("fill", "transparent")
          .attr("marker-end", "url(#end" + packet.id + ")");
      } else {
        // straight arrow: src -> dst
        svg.append("line")
          .style("stroke", colorScaleSIP.current(packet.branch))
          .attr("x1", srcArrowX)
          .attr("y1", arrowY)
          .attr("x2", dstArrowX)
          .attr("y2", arrowY)
          .attr("marker-end", "url(#end" + packet.id + ")");
      }

      // Create tooltips for each record
      select("#diagram").append("div")
        .attr("class", "tooltipDiagram tooltiptextCSS")
        .attr("id", "tooltip" + i)
        .style("display", "none")
        .style("opacity", 0);
      dragElement(document.getElementById("tooltip" + i));

      // Draw message labels
      const labelOffset = ((dstIndex - srcIndex) * VERT_SPACE) / 2;
      const labelX = XPAD + MESSAGE_LABEL_X_OFFSET + srcIndex * VERT_SPACE +
        labelOffset;
      const labelY = MESSAGE_LABEL_Y_OFFSET + i * MESSAGE_SPACE;

      svg.append("g")
        .attr("transform", `translate(${labelX}, ${labelY})`)
        .append("text")
        .attr("dx", "5px")
        .attr("dy", "-2px")
        .attr("text-anchor", "begin")
        .style("cursor", "grab")
        .style("font-size", "10px")
        .text(packet.method)
        .on("click", function (event) {
          // set z-index to front when you click on popup
          selectAll(".tooltipDiagram").style("z-index", 10);
          select("#tooltip" + i)
            .transition()
            .duration(200)
            .style("display", "inline-block")
            .attr("clicked", "true")
            .style("z-index", 20)
            .style("opacity", 1);

          let top = event.clientY;
          const maxY = document.documentElement.scrollHeight;
          const y = event.clientY;
          const x = event.clientX;
          if (maxY - top < 150) top = maxY - 200;

          select("#tooltip" + i).html(`
            <div class="tooltipDiagramHeader">
              <span style="cursor: default; margin-right: 10px; 
                margin-left: 3px; font-size: 12px; color: black;" 
                onclick=getElementById("${"tooltip" + i}").style.display="none">
                X
              </span>
              ${packet.method}
            </div>
            <div class="tooltipDiagramBody">
              ${formatSIPPayload(packet)}
            </div>`)
            .style("left", x + "px")
            .style("top", y + "px");
        });
    }

    // Arrow style
    svg.append("svg:defs").selectAll(".arrows")
      .data(sipPackets)
      .enter()
      .append("svg:marker")
      .attr("id", (d) => "end" + d.id)
      .attr("viewBox", "0 -5 10 10")
      .attr("refX", 10)
      .attr("refY", 0)
      .attr("markerWidth", 6)
      .attr("markerHeight", 6)
      .attr("orient", "auto")
      .style("fill", (d) => colorScaleSIP.current(d.branch))
      .append("svg:path")
      .attr("d", "M0,-5L10,0L0,5");

    function dragElement(element: HTMLElement | null) {
      if (!element) return;
      let mouseOffsetX = 0;
      let mouseOffsetY = 0;

      const header = document.getElementById(element.id + "Header");
      if (header) header.onmousedown = dragMouseDown;
      else element.onmousedown = dragMouseDown;

      function dragMouseDown(e: MouseEvent) {
        if (!element) return;
        const path = e.composedPath();
        const firstEl = path.at(0) as HTMLElement;
        if (
          firstEl.getAttribute("class") === "tooltipDiagramHeader"
        ) {
          e = e || window.event;
          e.preventDefault();

          const elRect = element.getBoundingClientRect();

          // get the mouse cursor offset to the element at startup:
          mouseOffsetX = e.clientX - elRect.x;
          mouseOffsetY = e.clientY - elRect.y;

          document.onmouseup = closeDragElement;
          document.onmousemove = elementDrag;
        }

        //set z-index to front when you click on popup
        const allTooltips = document.getElementsByClassName(
          "tooltipDiagram",
        ) as HTMLCollectionOf<HTMLElement>;
        for (const tooltip of allTooltips) {
          tooltip.style.zIndex = "10";
        }
        const firstElParent = firstEl.parentElement;
        if (firstElParent) firstElParent.style.zIndex = "20";

        function elementDrag(e: MouseEvent) {
          if (!element) return;
          e = e || window.event;
          e.preventDefault();

          // calculate the new element position:
          const elTop = e.clientY - mouseOffsetY;
          const elLeft = e.clientX - mouseOffsetX;

          // set the element's new position:
          element.style.top = elTop + "px";
          element.style.left = elLeft + "px";
        }
      }

      function closeDragElement() {
        // stop moving when mouse button is released:
        document.onmouseup = null;
        document.onmousemove = null;
      }
    }
  };

  return (
    <span
      style={{ width: "100%", marginLeft: "30px", marginTop: "20px" }}
    >
      <span>
        {download &&
          (
            <>
              <button
                className="noFormatButton"
                key="downloadPCAP"
                onClick={getPcap}
              >
                <img
                  className="icon"
                  alt="downloadIcon"
                  src={downloadPcapIcon}
                  title="download PCAP"
                />
              </button>
              <button
                className="noFormatButton"
                key="downloadSD"
                onClick={getDiagram}
              >
                <img
                  className="icon"
                  alt="downloadIcon"
                  src={downloadIcon}
                  title="download chart"
                />
              </button>
            </>
          )}
        {[...data?.sipFlow.hosts ?? []].map((name) => (
          <span style={{ "marginLeft": "5px" }} key={name}>
            <input
              type="checkbox"
              defaultChecked
              id={name}
              onClick={updateSelection}
              style={{ "verticalAlign": "middle" }}
            >
            </input>
            <label
              style={{ "marginLeft": "5px", "verticalAlign": "middle" }}
            >
              {name}
            </label>
          </span>
        ))}
      </span>
      <div
        style={{ width: "100%", display: "flex", justifyContent: "center" }}
        id="diagram"
      >
        {error || noData &&
            (
              <div
                style={{
                  "textAlign": "center",
                  "margin": "auto",
                  "marginTop": "10%",
                  "marginLeft": "80px",
                }}
              >
                {noData &&
                  <b>Trying to get pcap file...</b>}
                {error &&
                  (
                    <>
                      <b>Whoops</b>
                      <p>{error}</p>
                    </>
                  )}
              </div>
            )}
      </div>
    </span>
  );
}
