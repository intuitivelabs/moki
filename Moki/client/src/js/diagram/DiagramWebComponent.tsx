import r2wc from "@r2wc/react-to-web-component"
import { SequenceDiagramRender } from "./sequenceDiagram";

function DiagramWebComponent({ content }: { content: string }) {
  return (
      <SequenceDiagramRender
        {...{
          download: false,
          load: async () => {
            return content;
          },
          filenames: undefined,
        }}
      />
  );
}

const Diagram = r2wc(DiagramWebComponent, {
  props: { content: "string" },
});
customElements.define("sequence-diagram", Diagram);
