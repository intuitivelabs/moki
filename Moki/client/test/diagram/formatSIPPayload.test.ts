import { formatSIPPayload } from "@/js/diagram/formatPayloadSIP";
import { PacketSIP } from "@/js/diagram/parseFlow";
import { expect, it } from "vitest";

it("Format SIP Payload", () => {
  const packet = {
    method: "INVITE",
    payload: `
Call-ID: bliup@bloup
CSeq: 213 INVITE
Content-Length: 232

o=bloup
c=bla
awdc=bloup
m=audio`,
  } as PacketSIP;
  expect(formatSIPPayload(packet)).toMatchSnapshot()
});
