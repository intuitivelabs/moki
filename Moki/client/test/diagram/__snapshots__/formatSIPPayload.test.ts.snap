// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Format SIP Payload 1`] = `
"<div><b>INVITE</b></div><div>
        <span className=\\"key\\"><b>Call-ID: </b></span>
        <span className=\\"value\\"> bliup@bloup</span>
      </div><div>
        <span className=\\"key\\"><b>CSeq: </b></span>
        <span className=\\"value\\"> 213 INVITE</span>
      </div><div>
        <span className=\\"key\\"><b>Content-Length: </b></span>
        <span className=\\"value\\"> 232</span>
      </div><br /><div>
      <span className=\\"value\\" style=color:green>
        c=bla
      </span></div><div>
      <span className=\\"value\\" style=>
        awdc=bloup
      </span></div><div>
      <span className=\\"value\\" style=color:green>
        m=audio
      </span></div>"
`;
